protos/flyteidl/datacatalog/datacatalog.proto

```diff
@@ -150,6 +173,47 @@ message ListDatasetsResponse {
     string next_token = 2;
 }
 
+// Get the Artifact or try to reserve a spot if the Artifact does not exist.
+message GetOrReserveArtifactRequest {
+    DatasetID dataset_id = 1;
+    string tag_name = 2;
+    string owner_id = 3;
```

Comment:
I dropped the timeout from request here since it will be a global timeout and I think we can make it configurable via some config file in datacatalog.